### PR TITLE
Make receival of a known block trigger a refetch of missing dependencies 

### PR DESCRIPTION
**Pull request by KentShikama**
### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
Add link to corresponding JIRA issue.

https://rchain.atlassian.net/browse/RHOL-975

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 1]
```diff
@@ -108,6 +108,29 @@ class MultiParentCasperImpl[F[_]: Sync: Capture: ConnectionsCell: TransportLayer
...
+          dag.dataLookup.contains(blockHash) || blockBuffer.exists(_.blockHash == blockHash)
+      )
+      _ <- missingDependencies.traverse(hash => handleMissingDependency(hash, b))
+      possiblyFailedDependencyFetches = allDependencies.foldLeft(Set.empty[BlockMessage]) {
+        case (acc, blockHash) =>
+          blockBuffer.find(_.blockHash == blockHash) match {
+            case Some(block) => acc + block
+            case None        => acc
+          }
+      }
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala]: The computation of possiblyFailedDependencyFetches can be written in a more readable fashion
```      possiblyFailedDependencyFetches = 
        allDependencies.flatMap(blockHash => blockBuffer.find(_.blockHash == blockHash))
          .foldLeft(Set.empty[BlockMessage]) {case (acc,blockMessage) => acc + blockMessage}
```

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 2]
```diff
@@ -353,6 +353,7 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
         isOldBlock <- MultiParentCasper[F].contains(b)
         _ <- if (isOldBlock) {
               Log[F].info(s"Received block ${PrettyPrinter.buildString(b.blockHash)} again.")
+              MultiParentCasper[F].refetchDependencies(b)
```
Comment (birchmd) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: Missing `*>` from previous effect (i.e. logging will not actually happen when `F` is not `Id`).

**Comment by KentShikama** [comment]
@birchmd PTAL

File: casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala [hunk 3]
```diff
@@ -89,6 +89,7 @@ private class SleepingMultiParentCasperImpl[F[_]: Monad: Time](underlying: Multi
   def lastFinalizedBlock: F[BlockMessage]          = underlying.lastFinalizedBlock
   def storageContents(hash: ByteString): F[String] = underlying.storageContents(hash)
   def getRuntimeManager: F[Option[RuntimeManager]] = underlying.getRuntimeManager
+  def fetchDependencies: F[Unit]                   = ().pure[F]
```
Comment (birchmd) [inline casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: should be `underlying.fetchDependencies`

File: node/src/main/scala/coop/rchain/node/node.scala [hunk 4]
```diff
@@ -291,6 +289,15 @@ class NodeRuntime(conf: Configuration, host: String, scheduler: Scheduler) {
       _ <- time.sleep(5.seconds).toEffect
     } yield ()
 
+    val casperLoop: Effect[Unit] =
+      for {
+        _ <- casperConstructor.get map {
+              case Some(casper) => casper.fetchDependencies
+              case None         => ().pure[Effect]
+            }
+        _ <- time.sleep(5.seconds).toEffect
```
Comment (birchmd) [inline node/src/main/scala/coop/rchain/node/node.scala]: 5 seconds seems pretty short to me. Maybe 30 seconds?

File: casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala [hunk 5]
```diff
@@ -756,15 +756,22 @@ class HashSetCasperTest extends FlatSpec with Matchers {
...
+    nodes(0).casperEff.addBlock(block11)
 
-    (0 to 10).foreach { i =>
+    // Cycle of requesting and passing blocks until block #3 from nodes(0) to nodes(1)
+    (0 to 8).foreach { i =>
       nodes(1).receive()
       nodes(0).receive()
     }
 
+    // We simulate a network failure here by not allowing block #2 to get passed to nodes(1)
```
Comment (birchmd) [inline casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala]: What is this comment referring to now?

Comment (KentShikama) [reply casper/src/test/scala/coop/rchain/casper/HashSetCasperTest.scala]: It still applies. Above we only allow until block #3 to get passed to nodes(1).

File: casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala (hunk 3 above)
Comment (KentShikama) [reply casper/src/test/scala/coop/rchain/casper/api/CreateBlockAPITest.scala]: Ah yes. Nice catch

File: node/src/main/scala/coop/rchain/node/node.scala (hunk 4 above)
Comment (KentShikama) [reply node/src/main/scala/coop/rchain/node/node.scala]: Ok

**Pull request merged**